ldp/alg/callbacks.py

```diff
@@ -123,54 +122,64 @@ async def after_update(self) -> None:
-        self.output_dir.mkdir(parents=True, exist_ok=True)
+    @staticmethod
+    def default_serialize_env(env: Environment, transition: Transition) -> str | None:
+        if not transition.done:
+            return None
+        try:
+            frame = env.export_frame()
+        except NotImplementedError:
+            return None  # Allow for envs that didn't implement export_frame()
+        return frame.model_dump_json(exclude={"state"}, indent=2)
 
-        self.out_files: dict[str, Path] = {}
+    def __init__(
+        self,
+        output_dir: str | os.PathLike | None = None,
+        serialize_env: Callable[
+            [Environment, Transition], str | None
+        ] = default_serialize_env,
+    ):
+        if output_dir is not None:
+            self.output_dir: Path | None = Path(output_dir)
+            self.output_dir.mkdir(parents=True, exist_ok=True)
+        else:
+            self.output_dir = None
+
+        self.traj_files: dict[str, Path] = {}
         self.env_files: dict[str, Path] = {}
-        self.trajs: dict[str, Trajectory] = defaultdict(Trajectory)
+        self.traj_id_to_traj: dict[str, Trajectory] = defaultdict(Trajectory)
+        self.train_traj_ids: list[str | None] = []
+        self.eval_traj_ids: list[str | None] = []
+        self.serialize_env = serialize_env
 
-    def _make_filename(self, traj_id: str, env: Environment) -> str:
+    def _make_traj_filename(self, traj_id: str, env: Environment) -> str:
         """Create the filename for the output file."""
         return f"{traj_id}.jsonl"
 
     async def after_transition(
         self, traj_id: str, agent: Agent, env: Environment, transition: Transition
     ) -> None:
-        if traj_id not in self.out_files:
-            self.out_files[traj_id] = self.output_dir / self._make_filename(
+        if self.output_dir is not None and traj_id not in self.traj_files:
+            self.traj_files[traj_id] = self.output_dir / self._make_traj_filename(
                 traj_id, env
             )
             self.env_files[traj_id] = self.output_dir / f"{traj_id}_env.json"
 
-        traj = self.trajs[traj_id]
+        traj = self.traj_id_to_traj[traj_id]  # Get or create trajectory
         traj.steps.append(transition)
-        # TODO: make this async?
-        traj.to_jsonl(self.out_files[traj_id])
-        if transition.done:
-            with (
-                # Do not fail if the environment didn't implement export_frame().
-                suppress(NotImplementedError),
-                Path(self.env_files[traj_id]).open("w") as f,
-            ):
-                f.write(env.export_frame().model_dump_json(exclude={"state"}, indent=2))
+        if self.output_dir is not None:
+            await traj.to_jsonl(self.traj_files[traj_id])
+            async with aiofiles.open(self.env_files[traj_id], "w") as f:
+                if to_dump := self.serialize_env(env, transition):
+                    await f.write(to_dump)
```

Comment:
Minor optimization, but perhaps could be `asyncio.gather`'d?

Reply:
Sure nice comment, just implemented this